src/PureScript/Backend/Chez/Convert.purs

```diff
@@ -137,28 +137,26 @@ codegenExpr codegenEnv@{ currentModule } s = case unwrap s of
     S.Identifier $ coerce $ S.toChezIdent i l
   Lit l ->
     codegenLiteral codegenEnv l
+
   App f p ->
     S.chezCurriedApplication (codegenExpr codegenEnv f) (codegenExpr codegenEnv <$> p)
   Abs a e -> do
     S.chezCurriedFunction (uncurry S.toChezIdent <$> a) (codegenExpr codegenEnv e)
+  UncurriedApp f p ->
+    S.chezUncurriedApplication (codegenExpr codegenEnv f) (codegenExpr codegenEnv <$> p)
+  UncurriedAbs a e ->
+    S.chezUncurriedFunction (uncurry S.toChezIdent <$> a) (codegenExpr codegenEnv e)
+  UncurriedEffectApp f p ->
+    S.chezThunk $ S.chezUncurriedApplication (codegenExpr codegenEnv f)
+      (codegenExpr codegenEnv <$> p)
+  UncurriedEffectAbs a e ->
+    S.chezUncurriedFunction (uncurry S.toChezIdent <$> a)
+      (S.chezUnthunk $ codegenPureChain codegenEnv e)
```

Comment:
Shouldn't these not be thunked?

Reply:
For `UncurriedEffectApp`, applying `f` to `p` runs the effect immediately without the `(lambda () ..)` thunking. Likewise for `UncurriedEffectAbs`, since `e` evaluates to a thunked value, it has to be unthunked.

_I think_ this can be replaced this way though, but I'm not entirely sure...
```suggestion
  UncurriedEffectApp f p ->
    S.chezThunk $ S.chezUncurriedApplication (codegenExpr codegenEnv f)
      (codegenExpr codegenEnv <$> p)
  UncurriedEffectAbs a e ->
    S.chezUncurriedFunction (uncurry S.toChezIdent <$> a)
      (codegenChain effectBlockMode codegenEnv e)
      -- using codegenEffectChain adds thunking here, which we don't want to do...
```

Reply:
Representationally, EffectFn should always be identicaly to an uncurried function. That is, it doesn't have any additional `(..)` invocation as an `Effect` does. For example, EffectFn2 it is not equivalent to `Fn2 ... (Effect a)`, it is equivalent to `Fn2 ... a`, just impure to the type system. `runEffectFn` function however, _do_ return an `Effect`. The reason for this is that, within an effect _block_ (that is, under an EffectBind), EffectFn functions do not have additional `()` overhead, so multiple calls to EffectFns can be invoked without overhead _in addition to_ the thunk created for the given block.

For example, in `backend-es`:

https://github.com/aristanetworks/purescript-backend-optimizer/blob/85a70a4376bd6b5ab7ae13a6ee1afc8974e06f9c/backend-es/src/PureScript/Backend/Optimizer/Codegen/EcmaScript/Convert.purs#L269-L278

https://github.com/aristanetworks/purescript-backend-optimizer/blob/85a70a4376bd6b5ab7ae13a6ee1afc8974e06f9c/backend-es/src/PureScript/Backend/Optimizer/Codegen/EcmaScript/Convert.purs#L396

https://github.com/aristanetworks/purescript-backend-optimizer/blob/blob/85a70a4376bd6b5ab7ae13a6ee1afc8974e06f9c/backend-es/src/PureScript/Backend/Optimizer/Codegen/EcmaScript/Convert.purs#L475-L476